lib/docker-delta.coffee

```diff
@@ -119,27 +119,37 @@ hardlinkCopy = (srcRoot, dstRoot, linkDests) ->
 	rsync.waitAsync()
 
 applyBatch = (rsync, batch, timeout) ->
-	new Promise (resolve) ->
-		batch.pipe(rsync.stdin).on('unpipe', resolve)
-	.then ->
-		if timeout is 0
-			# wait for clean exit
-			return rsync.waitAsync()
-		else
-			# or until the given timeout is exceeded
-			return rsync.waitAsync().timeout(timeout)
-	.catch Promise.TimeoutError, ->
-		# timed out; kill it.
+	p = new Promise (resolve, reject) ->
+		batch.pipe(rsync.stdin)
+		.on('error', reject)
+		.on('finish', resolve)
+
+	if timeout isnt 0
+		p = p.timeout(timeout)
+
+	return p.then ->
+		# `p` is resolved on 'finish' but that doesn't guarantee rsync exited cleanly :/
+		# normally we'd end up here on rsync exiting with 0, but there are cases
+		# our pipe to its stdin is broken with no error and rsync still is up.
+		# so to find the reason rsync exited, we "poll" it once here via `waitAsync`.
+		# if rsync exited cleanly, `waitAsync` will resolve and the chain continues.
+		# for any other case, a short timeout will ensure we don't wait forever and
+		# manually kill rsync below.
+		rsync.waitAsync().timeout(5000)
+	.catch (err) ->
 		rsync.kill('SIGUSR1')
-		rsync.waitAsync()
+		rsync.waitAsync().throw(err)
```

Comment:
This `waitAsync` has no timeout, what happens if the rsync process gets in a weird hung state where it ignores `SIGUSR1`?

Reply:
Well, the signal to kill rsync is `SIGUSR1` and if it doesn't respond there's nothing we can do -- it doesn't respond to any signal (not even `SIGKILL`, which is weird) if node exits but rsync is still up. In any case, I ran a few weird test cases (corrupt delta, wrong src, etc.) and docker-delta was always able to take down rsync before the whole process ends, so I'm pretty confident we won't see any issues like that.